src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/TypeLoaderExports.cs

```diff
@@ -28,345 +29,248 @@ public static unsafe void ActivatorCreateInstanceAny(ref object ptrToData, IntPt
+                            if (result == IntPtr.Zero)
+                                result = RuntimeAugments.GetFallbackDefaultConstructor();
+                            return result;
+                        });
+            }
+
+            RawCalliHelper.Call(v._result, ptrToData);
         }
 
         //
         // Generic lookup cache
         //
 
-        private class Entry
+        private struct Key : IEquatable<Key>
+        {
+            public IntPtr _context;
+            public IntPtr _signature;
+
+            public Key(nint context, nint signature)
+            {
+                _context = context;
+                _signature = signature;
+            }
+
+            public bool Equals(Key other)
+            {
+                return _context == other._context && _signature == other._signature;
+            }
+
+            public override int GetHashCode()
+            {
+                // pointers will likely match and cancel out in the upper bits
+                // we will rotate context by 16 bit to keep more varying bits in the hash
+                IntPtr context = (IntPtr)BitOperations.RotateLeft((nuint)_context, 16);
+                return (context ^ _signature).GetHashCode();
+            }
+
+            public override bool Equals(object obj)
+            {
+                return obj is Key && Equals((Key)obj);
+            }
+        }
+
+        private struct Value
         {
-            public IntPtr Context;
-            public IntPtr Signature;
-            public IntPtr Result;
-            public IntPtr AuxResult;
-            public Entry Next;
+            public IntPtr _result;
+            public IntPtr _auxResult;
+
+            public Value(IntPtr result, IntPtr auxResult)
+            {
+                _result = result;
+                _auxResult = auxResult;
+            }
         }
 
-        // Initialize the cache eagerly to avoid null checks.
-        // Use array with just single element to make this pay-for-play. The actual cache will be allocated only
-        // once the lazy lookups are actually needed.
-        private static Entry[] s_cache;
+        //
+        // Parameters and state used by generic lookup cache resizing algorithm
+        //
 
-        private static Lock s_lock;
-        private static GCHandle s_previousCache;
+#if DEBUG
+        // use smaller numbers to hit resizing/preempting logic in debug
+        private const int InitialCacheSize = 8; // MUST BE A POWER OF TWO
+        private const int MaximumCacheSize = 512;
+#else
+        private const int InitialCacheSize = 128; // MUST BE A POWER OF TWO
+        private const int MaximumCacheSize = 128 * 1024;
+#endif // DEBUG
 
+        // Initialize the cache eagerly to avoid null checks.
+        private static GenericCache<Key, Value> s_cache;
         internal static void Initialize()
```

Comment:
Use `EagerStaticClassConstruction` for this so that it has the same structure as cast cache?

`EagerStaticClassConstruction` should be better for trimming. The compiler should be able to trim the initialization if the type is not otherwise used.

Reply:
I did not know the differences. I just used what this cache used before assuming there was a reason to use the `Initialize` pattern. Makes sense to switch to `EagerStaticClassConstruction`.